### PR TITLE
CAPTCHA problem and anonymity

### PR DESCRIPTION
**Pull request by anuran-roy**
Added proxy modules (experimental) with google search results integration.

### CODE REVIEW

File: core/util/google.py [hunk 1]
```diff
@@ -17,16 +17,18 @@
 
 import re
 import os
-
+from core.util import proxy
```
Comment (saeeddhqan) [inline core/util/google.py]: self.framework.proxy?

File: core/util/google.py [hunk 2]
```diff
@@ -17,16 +17,18 @@
...
 class main:
 	# framework = None
-	def __init__(self, q, limit=1, count=10, google_api=None, google_cx=None):
+	countip=0
```
Comment (saeeddhqan) [inline core/util/google.py]: not here

File: core/util/google.py [hunk 3]
```diff
@@ -17,16 +17,18 @@
 
 import re
 import os
-
+from core.util import proxy
 class main:
 	# framework = None
```
Comment (saeeddhqan) [inline core/util/google.py]: remove it.

**Comment by rachejazz** [comment]
@anuran-roy the slack channels had some complaints that google is having captcha triggered always. Have you been facing any errrors?

File: core/util/google.py [hunk 4]
```diff
@@ -47,18 +50,29 @@ def run_crawl(self):
...
+						headers={'User-Agent': self.agent})	
 			except Exception as e:
 				self.framework.error(f"[GOOGLE] ConnectionError: {e}")
 				return
 			if req.status_code in (503, 429):
 				req = self.framework.error('[GOOGLE] Google CAPTCHA triggered.')
-				break
-
+				countip+=1
+				if(proxy.rotateip(countip)==-1):
```
Comment (saeeddhqan) [inline core/util/google.py]: if proxy.rotateip(countip) == -1:

File: core/util/google.py [hunk 5]
```diff
@@ -47,18 +50,29 @@ def run_crawl(self):
...
+					print('[PROXY] End of proxy list. ')
```
Comment (saeeddhqan) [inline core/util/google.py]: self.framework.output

File: core/util/proxy.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
+import requests
```
Comment (saeeddhqan) [inline core/util/proxy.py]: copyright comment

File: core/util/proxy.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        use : Determine whether to use proxy or not.
+        '''
+        self.use=use
+        self.verbose=verbose
+    iplist=[]
+    proxydict={}
+
+    def getproxy(self,use=False):
+        if self.use==True:
+            print("[PROXY] Gathering proxies. It's quite time consuming, so do some pushups in the meantime :P")
```
Comment (saeeddhqan) [inline core/util/proxy.py]: self.output

File: core/util/proxy.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        '''
+        use : Determine whether to use proxy or not.
+        '''
+        self.use=use
+        self.verbose=verbose
+    iplist=[]
+    proxydict={}
+
+    def getproxy(self,use=False):
+        if self.use==True:
```
Comment (saeeddhqan) [inline core/util/proxy.py]: spaces: if self.use == True:

File: core/util/proxy.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+class main:
+
+    def __init__(self, use=False, verbose=False):
+        '''
+        use : Determine whether to use proxy or not.
+        '''
+        self.use=use
+        self.verbose=verbose
+    iplist=[]
```
Comment (saeeddhqan) [inline core/util/proxy.py]: not here

File: core/util/proxy.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    proxydict={}
```
Comment (saeeddhqan) [inline core/util/proxy.py]: not here

File: core/util/proxy.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+    def getproxy(self,use=False):
+        if self.use==True:
+            print("[PROXY] Gathering proxies. It's quite time consuming, so do some pushups in the meantime :P")
+            URL=['https://premproxy.com/list/']
```
Comment (saeeddhqan) [inline core/util/proxy.py]: spaces

File: core/util/proxy.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+            i=2
```
Comment (saeeddhqan) [inline core/util/proxy.py]: spaces: i = 2
apply it for all

File: core/util/proxy.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+                for purl in URL:
+                    r=requests.get(purl)
+                    sp=BeautifulSoup(r.content,'html.parser')
+                    lst=list(sp.find_all('input',attrs={"name" : "proxyIp[]"}))
+                    for l in lst:
+                        ip=str(l.get('value'))[0:-6]
+                        ips.append(ip)
+                    time.sleep(2)
+                #print(ips[0:5])
+                print("[PROXY] Proxies gathered.")
```
Comment (saeeddhqan) [inline core/util/proxy.py]: self.framework.output

File: core/util/proxy.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+            except Exception as e1:
+                print(f"An error occured. Error details: {e1}")
```
Comment (saeeddhqan) [inline core/util/proxy.py]: self.framework.output for all. put a [PROXY] before the text.

File: core/util/proxy.py [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+                        ip=str(l.get('value'))[0:-6]
+                        ips.append(ip)
+                    time.sleep(2)
+                #print(ips[0:5])
+                print("[PROXY] Proxies gathered.")
+            except Exception as e1:
+                print(f"An error occured. Error details: {e1}")    
+            j=0    
+            while(j<len(ips)):
+                ips[j]="http://"+ips[j]
```
Comment (saeeddhqan) [inline core/util/proxy.py]: f"http://{ips[j]}"

File: core/util/proxy.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+                URL.append(purl)
+                i+=1    
+
+            ips=[]
+            ips2=[]
+            purl=URL[0]
+
+            try:
+                for purl in URL:
+                    r=requests.get(purl)
```
Comment (saeeddhqan) [inline core/util/proxy.py]: self.framework.request? so remove requests

**Comment by anuran-roy** [comment]
> @anuran-roy the slack channels had some complaints that google is having captcha triggered always. Have you been facing any errrors?

@rachejazz this is the exact reason that we're working on this module. But I guess it'll take some time to be merged to the main branch, mostly because there will be quite a few changes in the existing codebase. For now, as you and others suggested, we have to switch search engines.

File: core/util/proxy.py [hunk 17]
```diff
@@ -0,0 +1,108 @@
+import requests
+from bs4 import BeautifulSoup
+import time
+
+class main:
+
+    def __init__(self, use=False, verbose=False):
```
Comment (saeeddhqan) [inline core/util/proxy.py]: use framework options. Something like this? self.framework._global_options['autoproxy')

**Review by saeeddhqan — COMMENTED** [review]
most common errors: spaces, use print, request instead of self.framework.output and self.framework.request. 
Correct those and we're going to the next level.

File: core/util/google.py (hunk 2 above)
Comment (anuran-roy) [reply core/util/google.py]: Sir I didn't get what you meant here. 😅

Should I remove `countip=0`? If yes, how do I replace it? 😅
It'll be really helpful if you could tell me.

File: core/util/google.py (hunk 3 above)
Comment (anuran-roy) [reply core/util/google.py]: Done.

File: core/util/google.py (hunk 1 above)
Comment (anuran-roy) [reply core/util/google.py]: Done.

File: core/util/google.py (hunk 4 above)
Comment (anuran-roy) [reply core/util/google.py]: Done

File: core/util/google.py (hunk 5 above)
Comment (anuran-roy) [reply core/util/google.py]: Done.

File: core/util/proxy.py (hunk 13 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done

File: core/util/proxy.py (hunk 14 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done.

File: core/util/proxy.py (hunk 15 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done.

File: core/util/proxy.py (hunk 16 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done

File: core/util/proxy.py (hunk 17 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done.

File: core/util/proxy.py (hunk 7 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done.

File: core/util/proxy.py (hunk 8 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done.

File: core/util/proxy.py (hunk 9 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Sir I didn't quite get what you meant. 😅

Should I remove `iplist=[]` from here? If yes, where and how should I write it? As `self.iplist=[]`?

File: core/util/proxy.py (hunk 11 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done.

File: core/util/proxy.py (hunk 12 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done.

File: core/util/proxy.py (hunk 10 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Sir I didn't quite get what you meant. 😅

Should I remove `proxydict={}` from here? If yes, where and how should I write it? As `self.proxydict={}`?

File: core/util/proxy.py (hunk 6 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done.

File: core/util/google.py (hunk 2 above)
Comment (saeeddhqan) [reply core/util/google.py]: Define it in init

File: core/util/google.py [hunk 18]
```diff
@@ -17,16 +17,18 @@
 
 import re
 import os
-
+#from core.util import proxy
 class main:
 	# framework = None
-	def __init__(self, q, limit=1, count=10, google_api=None, google_cx=None):
+	countip=0
+	def __init__(self, q, limit=1, count=10, google_api=None, google_cx=None)#, autoproxy=False):
```
Comment (saeeddhqan) [inline core/util/google.py]: ?

File: core/util/google.py [hunk 19]
```diff
@@ -37,8 +39,9 @@ def __init__(self, q, limit=1, count=10, google_api=None, google_cx=None):
 		self.google_api = google_api
 		self.google_cx = google_cx
 		self._links = []
-
+		self.autoproxy=self._global_options_['autoproxy']
 	def run_crawl(self):
+		global countip
```
Comment (saeeddhqan) [inline core/util/google.py]: remove it. you have self.countip!

**Comment by anuran-roy** [comment]
@saeeddhqan After your suggestions, these are the states:

[Proxy Util class](https://github.com/anuran-roy/Maryam/blob/dev-pinterest-module/core/util/proxy.py)
[Google Search Util class](https://github.com/anuran-roy/Maryam/blob/dev-pinterest-module/core/util/google.py)

I didn't get what you told me to do with the variables `countip=0`, `iplist=[]`, and `proxydict={}` 😅

File: core/util/google.py [hunk 20]
```diff
@@ -47,18 +50,29 @@ def run_crawl(self):
...
+						params=payload,
+						headers={'User-Agent': self.agent})	
 			except Exception as e:
 				self.framework.error(f"[GOOGLE] ConnectionError: {e}")
 				return
 			if req.status_code in (503, 429):
 				req = self.framework.error('[GOOGLE] Google CAPTCHA triggered.')
-				break
-
+				countip+=1
```
Comment (saeeddhqan) [inline core/util/google.py]: if autoproxy is true

**Comment by saeeddhqan** [comment]
After all, is it accountable for captcha or should we use alternative search engines that use google?

**Comment by anuran-roy** [comment]
@saeeddhqan I think that this issue will remain. This is because even if we use alternative search engines that use Google, it will be rate-limited and will have the CAPTCHA Problem again. 

What's more, the CAPTCHA appears in less crawls in those Google derivatives than the original Google search page itself. :sweat_smile:

**Comment by z3y50n** [comment]
From my experience you can't just crawl google using proxies. They changed their policy recently and got very strict. They have an api for crawling their engine now but it requires a key and it is not free.

File: core/util/google.py (hunk 18 above)
Comment (anuran-roy) [reply core/util/google.py]: Done.

File: core/util/google.py (hunk 19 above)
Comment (anuran-roy) [reply core/util/google.py]: Done.

**Comment by anuran-roy** [comment]
@z3y50n I tried to implement the proxy module such that it can be used for all search engines with only minimal changes in the search engine util classes. Since Google is only one among them, I'm reasonably confident that others will work with the proxy module after some slight modifications.

File: core/util/google.py (hunk 20 above)
Comment (anuran-roy) [reply core/util/google.py]: Done.

File: core/util/google.py [hunk 21]
```diff
@@ -20,7 +20,6 @@
 #from core.util import proxy
 class main:
 	# framework = None
-	countip=0
 	def __init__(self, q, limit=1, count=10, google_api=None, google_cx=None)#, autoproxy=False):
```
Comment (k4u5h1k) [inline core/util/google.py]: Please always test your code before every push.
Missing the colon at the end.
After fixing that I got a ton of indentation errors.

Comment (anuran-roy) [reply core/util/google.py]: I am really sorry for the missing colon.

But for the code indentation, I am not finding errors yet. Can you please share some screenshots here @k4u5h1k ? It'll be a great help.

Comment (k4u5h1k) [reply core/util/google.py]: If it works fine push it. Could've been an error on my side.

**Comment by saeeddhqan** [comment]
Look at the amendments and apply these code-style for the rest.

File: core/util/google.py (hunk 2 above)
Comment (anuran-roy) [reply core/util/google.py]: Done.

**Review by anuran-roy — COMMENTED** [review]
Done.

File: core/util/proxy.py [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+#import self.framework.requests
+from bs4 import BeautifulSoup
+import time
+#Python proxy module. Contributed by: @anuran-roy under @saeeddhqan
+class main:
+
+    def __init__(self, verbose = False):
```
Comment (saeeddhqan) [inline core/util/proxy.py]: we don't need verbose. it's usefulness so delete it comprehensively.

File: core/util/proxy.py [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+        '''
+        use : Determine whether to use proxy or not.
```
Comment (saeeddhqan) [inline core/util/proxy.py]: remove the comment

File: core/util/proxy.py [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+        self.verbose = verbose
+        self.iplist = []
+
+    def getproxy(self):
+        if self.use == True:
+            self.framework.output("[PROXY] Gathering proxies. It's quite time consuming, so do some pushups in the meantime :P")
+            URL = [ 'https://premproxy.com/list/' ]
+            i = 2
+            while i<=8:
+                purl = 'https://premproxy.com/list/0'+str(i)+".htm"
```
Comment (saeeddhqan) [inline core/util/proxy.py]: : ), purl = f"....{i}.htm"

File: core/util/proxy.py [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+        self.use=self._global_options_['autoproxy']
+        self.verbose = verbose
+        self.iplist = []
+
+    def getproxy(self):
+        if self.use == True:
+            self.framework.output("[PROXY] Gathering proxies. It's quite time consuming, so do some pushups in the meantime :P")
+            URL = [ 'https://premproxy.com/list/' ]
+            i = 2
+            while i<=8:
```
Comment (saeeddhqan) [inline core/util/proxy.py]: while i <= 8:

File: core/util/proxy.py [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+        use : Determine whether to use proxy or not.
+        '''
+        self.use=self._global_options_['autoproxy']
+        self.verbose = verbose
+        self.iplist = []
+
+    def getproxy(self):
+        if self.use == True:
+            self.framework.output("[PROXY] Gathering proxies. It's quite time consuming, so do some pushups in the meantime :P")
+            URL = [ 'https://premproxy.com/list/' ]
```
Comment (saeeddhqan) [inline core/util/proxy.py]: don't use uppercase

File: core/util/proxy.py [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+            i = 2
+            while i<=8:
+                purl = 'https://premproxy.com/list/0'+str(i)+".htm"
+                URL.append(purl)
+                i += 1    
+            ips = []
+            ips2 = []
+            purl = URL[0]
+            try:
```
Comment (saeeddhqan) [inline core/util/proxy.py:58]: hard-coding. find the piece of code that yields an error, put it in the try block. rest should be in the else block.

File: core/util/proxy.py [hunk 28]
```diff
@@ -0,0 +1,115 @@
...
+                    for l in lst:
+                        ip = str(l.get('value'))[0:-6]
+                        ips.append(ip)
+                    time.sleep(2)
+                #self.framework.output(ips[0:5])
+                self.framework.output("[PROXY] Proxies gathered.")
+            except Exception as e1:
+                self.framework.error(f"[PROXY] An error occured. Error details: {e1}")    
+            j = 0
+            while(j<len(ips)):
```
Comment (saeeddhqan) [inline core/util/proxy.py]: while j < len(ips):

File: core/util/proxy.py [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+                ips[j] = f"http://{ips[j]}"
+                j += 1
+            j = 0
+            self.framework.output("[PROXY] Now checking for availability of the proxies...")
```
Comment (saeeddhqan) [inline core/util/proxy.py]: simple quote: '

File: core/util/proxy.py [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+            while(j<len(ips)):
```
Comment (saeeddhqan) [inline core/util/proxy.py]: while j < len(ips):

File: core/util/proxy.py [hunk 31]
```diff
@@ -0,0 +1,115 @@
...
+                try:
```
Comment (saeeddhqan) [inline core/util/proxy.py]: hard-coding

File: core/util/proxy.py [hunk 32]
```diff
@@ -0,0 +1,115 @@
...
+                    if int(self.framework.requests.get(ips[j]).status_code)==200:
```
Comment (saeeddhqan) [inline core/util/proxy.py]: bad idea. hard-coding. space errors.

File: core/util/proxy.py [hunk 33]
```diff
@@ -0,0 +1,115 @@
...
+                        ips2.append(ips[j])
+                    if self.verbose == True:
+                        self.framework.output(f"[PROXY] {ips[j]} is available! Yay! :D")
+                except Exception as e2:
```
Comment (saeeddhqan) [inline core/util/proxy.py]: bad idea. use self.framework.verbose

File: core/util/proxy.py [hunk 34]
```diff
@@ -0,0 +1,115 @@
...
+            while(j<len(ips)):
+                ips[j] = f"http://{ips[j]}"
+                j += 1
+            j = 0
+            self.framework.output("[PROXY] Now checking for availability of the proxies...")
+            while(j<len(ips)):
+                try:
+                    if int(self.framework.requests.get(ips[j]).status_code)==200:
+                        ips2.append(ips[j])
+                    if self.verbose == True:
```
Comment (saeeddhqan) [inline core/util/proxy.py]: bad idea.

File: core/util/proxy.py [hunk 35]
```diff
@@ -0,0 +1,115 @@
...
+                        self.framework.output(f"[PROXY] {ips[j]} is available! Yay! :D")
+                except Exception as e2:
+                    if self.verbose == True:
+                        self.framework.error(f"[PROXY] {ips[j]} is not available. :(")
+                        self.framework.error(f"[PROXY] An error occured. Error details: {e2}")
+                finally:
+                    j += 1 
+                ips = []                   
+            #self.framework.output(ips2)
+            self.framework.output("[PROXY] Availability check completed! Writing to file...")
```
Comment (saeeddhqan) [inline core/util/proxy.py]: quote error

**Review by saeeddhqan — COMMENTED** [review]
Your code style is not good. all of them need to rewrite. and I think we can just use it as a module that will find proxies?
What would you think?

**Comment by anuran-roy** [comment]
> Your code style is not good. all of them need to rewrite. and I think we can just use it as a module that will find proxies?
> What would you think?

Yes Sir, we can just keep it as a module. I'm completely fine with that. 😄 

And as for the rewriting of code, I'm working on it. 😃

**Comment by saeeddhqan** [comment]
Okay, thank you

**Comment by anuran-roy** [comment]
By the way, I noticed something today. Whenever we're calling to rotateip, it isn't working because we haven't fetched IPs from online. I think that we should do something to trigger the ip fetcher method whenever someone does `set autoproxy True`. This is probably the reason why google.py isn't working.

File: core/util/proxy.py (hunk 28 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done.

File: core/util/proxy.py (hunk 26 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done

File: core/util/proxy.py (hunk 25 above)
Comment (anuran-roy) [reply core/util/proxy.py]: I guess that this is important. Premproxy has only about 8 pages. Hence the limit @saeeddhqan . 😅

File: core/util/proxy.py (hunk 24 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done

File: core/util/proxy.py (hunk 23 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done

File: core/util/proxy.py (hunk 22 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done

File: core/util/proxy.py (hunk 31 above)
Comment (anuran-roy) [reply core/util/proxy.py]: I reused the variable. I guess that'd save a few extra bytes of memory from being used. 😅

File: core/util/proxy.py (hunk 35 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done.

File: core/util/proxy.py (hunk 34 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Removed verbose.

File: core/util/proxy.py (hunk 33 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Removed verbose.

File: core/util/proxy.py (hunk 32 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Spacing errors corrected. 

I don't get it 😅 The HTTP status code of OK has to be hardcoded, doesn't it? Or is there any other way to implement this?

File: core/util/proxy.py (hunk 29 above)
Comment (anuran-roy) [reply core/util/proxy.py]: Done.

File: core/util/proxy.py (hunk 27 above)
Comment (anuran-roy) [reply core/util/proxy.py:58]: I am reusing the variable, so that's why I am reinitializing them. Will save a few bytes of memory I guess 😅

**Review by anuran-roy — COMMENTED** [review]
> By the way, I noticed something today. Whenever we're calling to rotateip, it isn't working because we haven't fetched IPs from online. I think that we should do something to trigger the ip fetcher method whenever someone does `set autoproxy True`. This is probably the reason why google.py isn't working.

This is regarding the modified google.py script, which doesn't fetch any results from social media. This doesn not show any error - it simply shows no results. @k4u5h1k can you please verify it? 😅

**Comment by k4u5h1k** [comment]
It does give an error
Screenshot 2021-04-08 at 10 30 49 AM   
change `self._global_options_` to `self.framework._global_options`   

After fixing that I am getting  
`[!] GoogleCodeError: getproxy() missing 1 required positional argument: 'self'.`
All the best

File: core/util/google.py [hunk 36]
```diff
@@ -37,7 +38,10 @@ def __init__(self, q, limit=1, count=10, google_api=None, google_cx=None):
 		self.google_api = google_api
 		self.google_cx = google_cx
 		self._links = []
-
+		self.autoproxy = self._global_options_['autoproxy']
```
Comment (saeeddhqan) [inline core/util/google.py]: it gives error

**Comment by anuran-roy** [comment]
Seems like this is an issue that will take quite some time to resolve. I'll try my best to resolve it as soon as possible, but it seems that it will take some time. 😞

**Comment by anuran-roy** [comment]
I'm really sorry for my inactivity for a few days... Really really sorry for that. I was unwell.

BTW, I am stuck with a small issue... Does Maryam have a self.framework.request implementation of `requests.get`? Because when I did `self.framework.request.get`, it threw up an error.

**Comment by z3y50n** [comment]
You use `self.framework.request` and specify the method as a parameter.

You can check how the function is defined here:

https://github.com/saeeddhqan/Maryam/blob/f938913d962b23f5090d6c88553a87eaeda29a39/core/core.py#L481-L515

**Comment by anuran-roy** [comment]
Finally google.py and proxy.py are working properly. But I have a small issue. It is regarding setting autoproxy on or off. For now, I have set them to true. I have commented out what I wrote when I tried to fetch the autoproxy permissions, so you can see what I tried actually, and tell me what changes to make.

Working screenshot:

image

This is the first try. Almost immediately afterwards, I tried it again. In normal case, I would've got an error `Google CAPTCHA Triggered`

Here's the second try:

image

@saeeddhqan @k4u5h1k @rachejazz I think there are some code style inconsistencies, so I'll be working towards improving them. 😅

**Comment by saeeddhqan** [comment]
Check with a big doze as I told you previously.

**Comment by anuran-roy** [comment]
Yeah sure! I'll work on them as soon as I can!

**Comment by anuran-roy** [comment]
Okay, I fixed most of the coding inconsistencies I guess. Only issue remaining for now is in the environment variable I guess. Autoproxy is getting triggered even if I am not mentioning `set autoproxy True`

That's all I guess. 😅 

@saeeddhqan can you please see it?

**Comment by saeeddhqan** [comment]
first: Check with a big doze as I told you previously.
then, show the results with estimated time.

**Comment by anuran-roy** [comment]
After hammering Google search results non-stop, I got the trigger that `Google CAPTCHA Triggered`. But then, I get an error from the document parser that the document to parse it is NIL. Even though I gave the `continue` statement, it doesn't seem to stop the page from being loaded into the pages list.

Screenshots:

1. image

(Error)

2.  
image

(Concerned logical block of code)

@saeeddhqan @k4u5h1k @z3y50n @rachejazz any idea how to overcome this?

**Pull request closed without merging**